File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.cpp
#include<string>
#include <iostream>
#include <queue>

using namespace std;

bool solution(string s)
{
    bool answer = true;
    
    queue<char> q;
    
    for (auto chr : s) {
        // 시작부터 이상할 경우
        if (chr == ')' && q.empty()) {
            return false;
        }
        
        // 짝을 맞을 때마다, 제거해주기
        if (chr == '(') {
            q.push(chr);
        } else {
            q.pop();
        }
    }
    
    if (q.empty()) {
        answer = true;
    } else {
        answer = false;
    }
    
    return answer;
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.py
def solution(s):
    answer = True
    q = []
    
    for i in s:
        if len(q) == 0 and i == ')':
            answer = False
            return answer
        
        if i == '(':
            q.append(i)
            
        if i == ')':
            q.pop()
            
    if len(q) > 0:
        answer = False
    else:
        answer = True

    return answer